tests/suites/test_suite_ssl.function

```diff
@@ -1755,3 +1981,76 @@ void ssl_session_serialize_version_check( int corrupt_major,
 
 }
 /* END_CASE */
+
+/* BEGIN_DEPENDENCIES
+ * depends_on:MBEDTLS_X509_CRT_PARSE_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_RSA_C:MBEDTLS_ECDSA_C
+ * END_DEPENDENCIES
+ */
+
+/* BEGIN_CASE */
+void test_handshake_state()
+{
+    enum { BUFFSIZE = 1024 };
+    int ret = 0;
+    mbedtls_endpoint ep;
+    mbedtls_endpoint_certificate cert;
+    mbedtls_mock_socket socket; /* dummy socket */
+
+    /* Test basic server initializing */
+
+    TEST_ASSERT( ( ret = mbedtls_endpoint_init( &ep, MBEDTLS_SSL_IS_SERVER ) )
+                 == 0 );
+    TEST_ASSERT( ( ret = mbedtls_endpoint_certificate_init( &ep, &cert ) )
+                 == 0 );
+    TEST_ASSERT( mbedtls_mock_socket_connect( &(ep.socket), &socket, BUFFSIZE )
+                 == 0 );
+
+    /* Test diffrent state of handshake for server */
+
+    TEST_ASSERT( ep.ssl.state == MBEDTLS_SSL_HELLO_REQUEST );
+    TEST_ASSERT( mbedtls_ssl_handshake_step( &(ep.ssl) ) == 0 );
+    TEST_ASSERT( ep.ssl.state == MBEDTLS_SSL_CLIENT_HELLO );
+
+    TEST_ASSERT( mbedtls_endpoint_reinit( &ep, MBEDTLS_SSL_IS_SERVER ) == 0 );
```

Comment:
Instead of reinitializing the endpoint, could we instead pass in parameters to this function?

This function would then be: init, connect, do some action (possibly selected by a `switch` with fallthrough; specified by parameter to this function), check result is as expected (expected result passed in as a parameter to this function).

This would make each test run more specific and therefore easier to debug. Good test cases should aim to have a singular purpose.

This may not be the best way, but here is what I mean by a `switch` with fallthrough. Feel free to do something different:

```C
switch (desired_state) {
case MBEDTLS_SSL_CLIENT_HELLO:
    TEST_ASSERT( mbedtls_ssl_handshake_step( &(ep.ssl) ) != 0 );
    /* fallthrough */
case MBEDTLS_SSL_HELLO_REQUEST:
    break;
}

/* We've now reached our desired state. Do state validation here. */
```

Reply:
Do you thing about one more universal function which will be run with different parameters from *.data file? For example 'test_handshake_state:MBEDTLS_SSL_SERVER_NEW_SESSION_TICKET:MBEDTLS_ERR_###'.
I can do that. Then we will be sure that testing different states will be independent of each other

Reply:
Yes, that's what I had in mind.